### PR TITLE
Email: Resend Token

### LINKED ISSUE #177: Add a link to resend email tokens
Most users that had enabled email tokens had a token already generated and stored which is why it now fails to send it after #156 was merged in.

The temporary solution is to add a "Re-send token" link.

### PR DESCRIPTION
**Pull request by kasparsd**
Fixes #177 which was introduced in #160, #155.

### CODE REVIEW

File: class.two-factor-core.php [hunk 1]
```diff
@@ -495,15 +495,27 @@ public static function login_form_validate_2fa() {
 			$provider = self::get_primary_provider_for_user( $user->ID );
 		}
 
+		// Allow providers to re-send codes, etc.
+		if ( $provider->process_request( $user ) ) {
```
Comment (kasparsd) [inline class.two-factor-core.php]: We introduce `process_request()` for when we don't want to process the actual 2FA. However, we do this only after verifying the 2FA nonce.

File: providers/class.two-factor-email.php [hunk 2]
```diff
@@ -152,6 +170,9 @@ public function authentication_page( $user ) {
 			<label for="authcode"><?php esc_html_e( 'Verification Code:', 'two-factor' ); ?></label>
 			<input type="tel" name="two-factor-email-code" id="authcode" class="input" value="" size="20" pattern="[0-9]*" />
 		</p>
+		<p class="two-factor-email-resend">
+			<input type="submit" class="button" name="two-factor-email-code-resend" value="<?php esc_attr_e( 'Resend Code', 'two-factor' ); ?>" />
```
Comment (kasparsd) [inline providers/class.two-factor-email.php]: I would prefer to have this as a regular link and a GET request instead but we don't have access to the current 2FA request token.

File: providers/class.two-factor-email.php [hunk 3]
```diff
@@ -167,6 +188,22 @@ public function authentication_page( $user ) {
...
 
 	/**
+	 * Send the email code, if missing or requested.
+	 *
+	 * @param  WP_USer $user WP_User object of the logged-in user.
+	 *
+	 * @return boolean       If the two factor should be authenticated.
+	 */
+	public function process_request( $user ) {
+		if ( isset( $user->ID ) && isset( $_POST['two-factor-email-code-resend'] ) ) {
```
Comment (kasparsd) [inline providers/class.two-factor-email.php]: Note that the 2FA request nonce has been already verified before this is called.

File: providers/class.two-factor-email.php [hunk 4]
```diff
@@ -79,6 +79,23 @@ public function user_has_token( $user_id ) {
...
+	 * Get the authentication token for the user.
+	 *
+	 * @param  int $user_id    User ID.
+	 *
+	 * @return string|boolean  User token or `false` if no token found.
+	 */
+	public function get_user_token( $user_id ) {
+		$hashed_token = get_user_meta( $user_id, self::TOKEN_META_KEY, true );
+
+		if ( ! empty( $hashed_token ) && is_string( $hashed_token ) ) {
```
Comment (georgestephanis) [inline providers/class.two-factor-email.php]: With `$single` being specified as `true` is there any way the returned data won't be a string?  Even if it's a number, wouldn't it be cast as a string?

File: providers/class.two-factor-provider.php [hunk 5]
```diff
@@ -45,6 +45,18 @@ public function print_label() {
...
 
 	/**
+	 * Allow providers to process the two factor authentication request for
+	 * additional tasks such as re-sending tokens, etc.
+	 *
+	 * @param WP_User $user WP_User object of the logged-in user.
+	 *
+	 * @return boolean Return `true` to prevent further processing.
+	 */
+	public function process_request( $user ) {
```
Comment (georgestephanis) [inline providers/class.two-factor-provider.php]: Should the function name here be `pre_process_request` or the like to acknowledge it runs first and can short out the other processing?

File: .travis.yml [hunk 6]
```diff
@@ -5,8 +5,6 @@ language:
     - node_js
 
 php:
-    - 5.2
```
Comment (kasparsd) [inline .travis.yml]: These are no longer available on the Trusty boxes.

File: providers/class.two-factor-email.php (hunk 4 above)
Comment (kasparsd) [reply providers/class.two-factor-email.php]: @georgestephanis This is mainly to protect from returning anything but a string or `false`.

File: tests/providers/class.two-factor-email.php [hunk 7]
```diff
@@ -169,4 +169,36 @@ function test_is_available_for_user() {
 		$this->assertTrue( $this->provider->is_available_for_user( false ) );
 	}
 
-}
\ No newline at end of file
+	/**
```
Comment (kasparsd) [inline tests/providers/class.two-factor-email.php]: And here some tests for the new methods.

**Pull request merged**